src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.CSharp.Tests.cs

```diff
@@ -0,0 +1,488 @@
+        }
+
+        [Fact]
+        public Task Diagnostic_StartsWith_ToUpperInvariant_If()
+        {
+            string originalCode = @"using System;
+class C
+{
+    int M()
+    {
+        string a = ""aBc"";
+        string b = ""bc"";
+        if (a.ToUpperInvariant().StartsWith(b))
+        {
+            return 5;
+        }
+        return 4;
+    }
+}";
+            string fixedCode = @"using System;
+class C
+{
+    int M()
+    {
+        string a = ""aBc"";
+        string b = ""bc"";
+        if (a.StartsWith(b, StringComparison.InvariantCultureIgnoreCase))
+        {
+            return 5;
+        }
+        return 4;
+    }
+}";
+            return VerifyCSharpAsync(originalCode, fixedCode, 8, 13, 8, 47, StringComparisonRule, StartsWithName);
+        }
+
+        [Fact]
+        public Task Diagnostic_CompareTo_ToLower_Assign()
+        {
+            string originalCode = @"using System;
+class C
+{
+    int M()
+    {
+        string a = ""aBc"";
+        string b = ""bc"";
+        int r = a.ToLower().CompareTo(b);
+        return r;
+    }
+}";
+            string fixedCode = @"using System;
+class C
+{
+    int M()
+    {
+        string a = ""aBc"";
+        string b = ""bc"";
+        int r = StringComparer.CurrentCultureIgnoreCase.Compare(a, b);
+        return r;
+    }
+}";
+            return VerifyCSharpAsync(originalCode, fixedCode, 8, 17, 8, 41, StringComparerRule, CompareToName);
+        }
+
+        [Fact]
+        public Task Diagnostic_CompareTo_ToLowerInvariant_Assign()
+        {
+            string originalCode = @"using System;
+class C
+{
+    int M()
+    {
+        string a = ""aBc"";
+        string b = ""bc"";
+        int r = a.ToLowerInvariant().CompareTo(b);
+        return r;
+    }
+}";
+            string fixedCode = @"using System;
+class C
+{
+    int M()
+    {
+        string a = ""aBc"";
+        string b = ""bc"";
+        int r = StringComparer.InvariantCultureIgnoreCase.Compare(a, b);
+        return r;
+    }
+}";
+            return VerifyCSharpAsync(originalCode, fixedCode, 8, 17, 8, 50, StringComparerRule, CompareToName);
+        }
+
+        [Fact]
+        public Task Diagnostic_CompareTo_ToUpper_Assign()
+        {
+            string originalCode = @"using System;
+class C
+{
+    int M()
+    {
+        string a = ""aBc"";
+        string b = ""bc"";
+        int r = a.ToUpper().CompareTo(b);
+        return r;
+    }
+}";
+            string fixedCode = @"using System;
+class C
+{
+    int M()
+    {
+        string a = ""aBc"";
+        string b = ""bc"";
+        int r = StringComparer.CurrentCultureIgnoreCase.Compare(a, b);
+        return r;
+    }
+}";
+            return VerifyCSharpAsync(originalCode, fixedCode, 8, 17, 8, 41, StringComparerRule, CompareToName);
+        }
+
+        [Fact]
+        public Task Diagnostic_CompareTo_ToUpperInvariant_Assign()
```

Comment:
Usually, we make test methods `async` and await the task instead of returning it.

Reply:
There's no need for that if the only asynchronous item inside your method is the object that is being returned. You can defer the awaiting to whoever is calling your method.

If I was doing more awaits across my method, I would not have a choice but to make my method async.

Reply:
I'd like to double check with the expert. @stephentoub would you mind correcting me if what I said was not accurate?

Reply:
The reason to make the method non-async and return the task would be performance, saving an extra state machine.  That can be worthwhile in production code.  It is not worthwhile here in a test.  In particular, by not having the extra state machine, if the test fails and you get an exception, the exception stack trace wouldn't have this test method in the call stack.

Reply:
Thank you for the explanation. I'll await the methods here.